apm-agent-plugins/apm-vertx/apm-vertx-3_6-plugin/src/main/java/co/elastic/apm/agent/vertx_3_6/helper/VertxWebHelper.java

```diff
@@ -0,0 +1,102 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.vertx_3_6.helper;
+
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.impl.GlobalTracer;
+import co.elastic.apm.agent.impl.transaction.Transaction;
+import co.elastic.apm.agent.sdk.weakmap.WeakMapSupplier;
+import co.elastic.apm.agent.vertx.AbstractVertxWebHelper;
+import com.blogspot.mydailyjava.weaklockfree.WeakConcurrentMap;
+import io.vertx.core.http.HttpServerRequest;
+import io.vertx.ext.web.RoutingContext;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.lang.reflect.Field;
+
+public class VertxWebHelper extends AbstractVertxWebHelper {
+    private static final VertxWebHelper INSTANCE = new VertxWebHelper(GlobalTracer.requireTracerImpl());
+
+    static final WeakConcurrentMap<HttpServerRequest, Transaction> requestTransactionMap = WeakMapSupplier.createMap();
+
+    public static VertxWebHelper getInstance() {
+        return INSTANCE;
+    }
+
+    public VertxWebHelper(ElasticApmTracer tracer) {
+        super(tracer);
+    }
+
+    public Transaction startOrGetTransaction(@Nonnull HttpServerRequest httpServerRequest) {
+        Transaction transaction = super.startOrGetTransaction(httpServerRequest);
+
+        if (transaction != null) {
+            requestTransactionMap.put(httpServerRequest, transaction);
+            enrichRequest(httpServerRequest, transaction);
+        }
+
+        return transaction;
+    }
+
+    @Override
+    public Transaction setRouteBasedNameForCurrentTransaction(RoutingContext routingContext) {
+        Transaction transaction = tracer.currentTransaction();
+        if (transaction == null) {
+            HttpServerRequest request = unwrapHttpServerRequest(routingContext.request());
+            transaction = getTransactionForRequest(request);
+        }
+
+        if (transaction != null) {
+            setRouteBasedTransactionName(transaction, routingContext);
+        }
+
+        return transaction;
+    }
+
+    private HttpServerRequest unwrapHttpServerRequest(HttpServerRequest request) {
+        try {
+            // reflection needed to unwrap the original request
+            if (request.getClass().getName().equals("io.vertx.ext.web.impl.HttpServerRequestWrapper")) {
+                Field delegateField = request.getClass().getDeclaredField("delegate");
+                delegateField.setAccessible(true);
```

Comment:
That fails when module encapsulation is enforced and causes a warning on Java 9+. Any alternatives to that which don't require unwrapping?

Reply:
Solved this with additional instrumentation. Though it still feels like a hacky workaround.